src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter.cs

```diff
@@ -24,12 +36,65 @@ public void OnTempDataSaving(ITempDataDictionary tempData)
                     var originalValue = kvp.Value;
 
                     var newValue = property.GetValue(Subject);
-                    if (newValue != null && newValue != originalValue)
+                    if (newValue != null && !newValue.Equals(originalValue))
                     {
                         tempData[Prefix + property.Name] = newValue;
                     }
                 }
             }
         }
+
+        public void OnActionExecuting(ActionExecutingContext context)
+        {
+            Subject = context.Controller;
+            var tempData = _factory.GetTempData(context.HttpContext);
+            var properties = GetSubjectProperties();
+            OriginalValues = new Dictionary<PropertyInfo, object>();
+
+            foreach (var property in properties)
+            {
+                var value = tempData[Prefix + property.Name];
+
+                OriginalValues[property] = value;
+
+                // TODO: Clarify what behavior should be for null values here
+                if (value != null && property.PropertyType.IsAssignableFrom(value.GetType()))
+                {
+                    property.SetValue(Subject, value);
+                }
```

Comment:
I don't think we need to check this here. 

As far as null, you want to set null if the property can accept a null.